### DISCUSSION TITLE
Add ability to completely disable request throttling

### DISCUSSION THREAD

**Discussion by mvantellingen** [storyblok/monoblok#183] [Ideas]
[storyblokurl]: https://www.storyblok.com?utm_source=github.com&utm_medium=issue-template

When using the storyblok-js-client in a server environment with a bit of traffic the storyblok-js-client causes significant slow downs due to request throttling. It should be possible to completely disable request throttling for performance (so not only changing the rateLimit setting).

For now we have created a fork which removes this functionality, see https://www.npmjs.com/package/@labdigital/storyblok-js-client but I would love to help get this option into the official client.

Note that while debugging the performance issue it also seems that the client creates quite some number of promises for requests, which also has impact on performance.

**Comment by rickvandermey** [comment]
It does have a huge performance improvement! However when we try the fork, it doesn't render anything on the server anymore. Something to keep in mind

**Comment by alvarosabu** [comment]
Hi @mvantellingen thanks for the feat request. Let me bring it up internally and keep you updated.

**Comment by Jonne** [comment]
@alvarosabu Any update on this? I'm also runnin into performance limitations because of this. Thanks! 🙏

**Comment by alvarosabu** [comment]
Hi @Jonne we are currently working on it. We are discussing to add a patch instead of waiting for v7 to release this. Will keep you posted here

**Comment by demeyerthom** [comment]
@alvarosabu any movement on this? I see v7 has come out already, so it does not look like it was part of these changes?

**Comment by imranolas** [comment]
Thanks for the patience on this 🙏  We recognise it's been nearly a year since this was first raised, and we appreciate the fork you've created in the meantime.

We definitely want to solve this throttling issue, but we're navigating some competing internal commitments. We're exploring automated client generation using OpenAPI, with the hope that throttling can be handled natively through rate limit headers rather than client-side logic.

Your fork approach is exactly what we'd recommend for teams facing this performance bottleneck right now. We'll keep this discussion prioritised as we work through our roadmap.

**Comment by johndetlefs** [comment]
Does this relate to the hanging promises error that's occurring when adding Storyblok to Shopify Hydrogen?

https://github.com/storyblok/storyblok-js-client/issues/682#issuecomment-2782961659

I've been trying to figure out a way around this while still allowing preview editing, and it's causing lots of issues while working locally. Is this the same issue (so wait for the update), or should I log a separate issue referencing the old one?

**Reply by z0n** [reply]
There's now a ticket for this as we also ran into this issue (and spent a lot of time to debug it): https://github.com/storyblok/monoblok/issues/319

For now, we're using the fork mentioned in the OP like this:

```ts
import StoryblokClient, { ISbConfig } from '@labdigital/storyblok-js-client';
import type { SbSDKOptions } from '@storyblok/js';
import { storyblokInit } from '@storyblok/react';

import { NotImplemented } from '~/components/storyblok/NotImplemented';
import { StoryblokPage } from '~/components/storyblok/StoryblokPage';

interface ApiPluginOptions extends SbSDKOptions {
  apiOptions?: ISbConfig;
}

// Implement a custom api plugin as the official one has a bug with rate limiting: https://github.com/storyblok/monoblok/issues/319
export const apiPlugin = (options: ApiPluginOptions = {}) => {
  const { apiOptions } = options;

  if (!apiOptions || !apiOptions.accessToken) {
    console.error(
      'You need to provide an access token to interact with Storyblok API. Read https://www.storyblok.com/docs/api/content-delivery#topics/authentication',
    );
    return;
  }

  const storyblokApi = new StoryblokClient(apiOptions);
  return { storyblokApi };
};

export const initStoryblok = () => {
  const components = {
    page: StoryblokPage,
  };

  const storyblokAccessToken =
    typeof import.meta.env.VITE_STORYBLOK_ACCESS_TOKEN === 'string'
      ? import.meta.env.VITE_STORYBLOK_ACCESS_TOKEN
      : '';

  storyblokInit({
    accessToken: storyblokAccessToken,
    use: [apiPlugin],
    bridge: true,
    components,
    customFallbackComponent: NotImplemented,
    enableFallbackComponent: true,
  });
};
```

And in the `root.tsx`, we're then simply calling `initStoryblok()`.


The apiPlugin function is basically just taken from here: https://github.com/storyblok/monoblok/blob/948ce49676e94dfcbb372f9007885045885f6954/packages/js/src/api.ts#L4-L16

**Comment by shabab477** [comment]
Any update on this ? We are running into a problem with throttling too but our needs are quite opposite. The issue is we have multiple build pipelines fetching from a single storyblok space consisting close to 1k stories as a statically built site working in parallel. Now having a delay of `0.3s` configured as const value is proving a problem by reaching the rate limit of the space. 

Having the rate limit configured as the max value does not help:

```
apiOptions: {
    rateLimit: 1000,
},
```

It would be helpful if something like @mvantellingen solution is added to the repo as done here: 
https://github.com/labd/storyblok-js-client/blob/92caa6cdc103f09c8702ab1d198e965d53bffcff/src/index.ts#L134

This is a good starting point for our use case but could be even nicer if the client delay function adds a jitter: 
https://github.com/storyblok/monoblok/blob/f3af69e744458cfe1f2834939e5afa4204bdf2ea/packages/js-client/src/utils.ts#L46-L48

**Comment by oliverstreissi** [comment]
This throttling issue caused a production outage for our Next.js App Router application. The problem took days to diagnose because the SDK provides no logging, metrics, or visibility into its queue behavior! 

The SDK conflicts with its own best practices: Storyblok's [official caching guide](https://www.storyblok.com/faq/how-stories-are-cached-content-delivery-api) recommends fetching the cache version (cv) from the `cdn/spaces/me` endpoint and including it in your content requests. This allows you to cache content permanently and only invalidate when the version changes. If you implement this pattern using the SDK for both the cv fetch and the content fetches, you create a circular dependency: your content requests need to wait for getCv() to complete, but getCv() itself uses the same SDK queue and must wait for available queue slots. Under concurrent load, the queue fills with content requests that are all waiting for getCv(), while getCv() waits for a queue slot that won't free up until those content requests complete. This creates a deadlock. Under load, this causes a complete application freeze. 

Apart from that, the SDK undermines Next.js App Router's design: Next.js automatically deduplicates identical fetch requests within a render. This is a core performance optimization. However, the SDK's throttle queue intercepts requests before fetch() is even called, so Next.js never gets a chance to deduplicate. A datasource  with translations called multiple times across components results in multiple separate throttled API calls when it should be a single request.

Why is there still no option to disable the SDKs throttling? Right now this is not usable in a server context without workarounds like using this fork [https://www.npmjs.com/package/@labdigital/storyblok-js-client](https://www.npmjs.com/package/@labdigital/storyblok-js-client).
